config/src/keys.rs

```diff
@@ -0,0 +1,236 @@
+    path::Path,
+};
+use toml;
+
+#[derive(Debug)]
+#[cfg_attr(any(test, feature = "testing"), derive(Clone))]
+enum PrivateKeyContainer<T> {
+    Present(T),
+    Removed,
+    Absent,
+}
+
+impl<T> PrivateKeyContainer<T>
+where
+    T: PrivateKey,
+{
+    pub fn take(&mut self) -> Option<T> {
+        match self {
+            PrivateKeyContainer::Present(_) => {
+                let key = std::mem::replace(self, PrivateKeyContainer::Removed);
+                match key {
+                    PrivateKeyContainer::Present(priv_key) => Some(priv_key),
+                    _ => unreachable!(),
+                }
+            }
+            _ => None,
+        }
+    }
+}
+
+impl<T> Serialize for PrivateKeyContainer<T>
+where
+    T: Serialize + ValidKeyStringExt,
+{
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            PrivateKeyContainer::Present(key) => serialize_key(key, serializer),
+            _ => serializer.serialize_str(""),
+        }
+    }
+}
+
+impl<'de, T> Deserialize<'de> for PrivateKeyContainer<T>
+where
+    T: ValidKeyStringExt + DeserializeOwned + 'static,
+{
+    fn deserialize<D>(deserializer: D) -> std::result::Result<PrivateKeyContainer<T>, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        // Note: Any error in parsing is assumed to be due to the private key being absent.
+        deserialize_key(deserializer)
+            .map(PrivateKeyContainer::Present)
+            .or_else(|_| Ok(PrivateKeyContainer::Absent))
+    }
+}
+
+// NetworkKeyPairs is used to store a node's Network specific keypairs.
+// It is filled via a config file at the moment.
+#[derive(Debug, Serialize, Deserialize)]
+#[cfg_attr(any(test, feature = "testing"), derive(Clone))]
+pub struct NetworkKeyPairs {
+    network_signing_private_key: PrivateKeyContainer<Ed25519PrivateKey>,
+    #[serde(serialize_with = "serialize_key")]
+    #[serde(deserialize_with = "deserialize_key")]
+    network_signing_public_key: Ed25519PublicKey,
+
+    #[serde(serialize_with = "serialize_key")]
+    #[serde(deserialize_with = "deserialize_key")]
+    network_identity_private_key: X25519StaticPrivateKey,
+    #[serde(serialize_with = "serialize_key")]
+    #[serde(deserialize_with = "deserialize_key")]
+    network_identity_public_key: X25519StaticPublicKey,
+}
+
+// required for serialization
+impl Default for NetworkKeyPairs {
+    fn default() -> Self {
+        let mut rng = StdRng::from_seed(TEST_SEED);
+        let (net_private_sig, net_public_sig) = compat::generate_keypair(&mut rng);
+        let (private_kex, public_kex) = x25519::compat::generate_keypair(&mut rng);
+        Self {
+            network_signing_private_key: PrivateKeyContainer::Present(net_private_sig),
+            network_signing_public_key: net_public_sig,
+            network_identity_private_key: private_kex,
+            network_identity_public_key: public_kex,
```

Comment:
public/private_kex => public/private_key

Reply:
name was added by @huitseeker, so I was going with the assumption that it was intentional to indicate that these are of type `x25519`.